# Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:
InvoiceNo – номер транзакции  
StockCode – код товара  
Description – описание товара  
Quantity – количество единиц товара, добавленных в заказ  
InvoiceDate – дата транзакции   
UnitPrice – цена за единицу товара  
CustomerID – id клиента  
Country – страна, где проживает клиент  



Импортируйте данные data.csv.zip (в папке) с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.

In [152]:
import pandas as pd
import os

In [153]:
path = r'lesson_5/data.csv.zip'

In [154]:
retail = pd.read_csv(path,compression='zip',encoding='ISO-8859-1')

In [155]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [156]:
retail_columns = retail.columns

# Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество. Если они есть, то удалите их из retail.

In [157]:
retail.duplicated().sum()

5268

In [158]:
print('В датафрейме встречается {} дубликатов.'.format(retail.duplicated().sum()))

В датафрейме встречается 5268 дубликатов.


In [159]:
retail.drop_duplicates(inplace=True)

# ⭐️Задание со звёздочкой!⭐️
Данные содержат в себе записи как и об успешных транзакциях, так и об отмененных. Если пользователь отменил заказ, в начале   номера транзакции (InvoiceNo) ставится C (canceled).   
Сколько всего заказов отменили пользователи?  

In [160]:
retail[retail['InvoiceNo'].str.startswith('C')].shape[0]

9251

In [161]:
print('Пользователи отменили {} заказов всего.'.format(retail[retail['InvoiceNo'].str.startswith('C')].shape[0]))

Пользователи отменили 9251 заказов всего.


# Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. В качестве ответа укажите число оставшихся строк.

In [162]:
retail = retail[retail['Quantity']>0]

In [163]:
print('Количество заказов, где количество единиц товара, добавленных в заказ больше нуля равно {}'.format(retail[retail['Quantity']>0].shape[0]))

Количество заказов, где количество единиц товара, добавленных в заказ больше нуля равно 526054


# 💀 Сложное задание! 💀
Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany).  
Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль.  
Запишите полученные id пользователей в germany_top (не весь датафрейм, только id).  
Идентификатор заказа – InvoiceNo.   
Для каждого заказа в данных может встречаться более 1 строки.

In [164]:
german_users = retail[retail['Country']=='Germany']

In [189]:
count_german_users = german_users.groupby('CustomerID',as_index=True) \
    .agg({'InvoiceNo':pd.Series.nunique}) \
    .sort_values('InvoiceNo',ascending=False) \
    .rename(columns={'InvoiceNo':'orders'})
count_german_users

orders
CustomerID        
12569.0         32
12471.0         30
12720.0         25
12621.0         20
12474.0         18
...            ...
12644.0          1
12612.0          1
12614.0          1
12622.0          1
12426.0          1

[94 rows x 1 columns]

In [190]:
# 80-ый процентиль:
N = count_german_users['orders'].quantile(0.8)
N

7.0

In [195]:
top_users = count_german_users[count_german_users['orders']>N].index

In [213]:
top_users

18

# ⭐️Задание со звёздочкой!⭐️
Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.


In [207]:
top_retail_germany = pd.DataFrame()
#top_retail_germany

In [208]:
for i in top_users:
    top_retail_germany = pd.concat((top_retail_germany,retail[retail['CustomerID']==i]))

In [215]:
top_retail_germany

InvoiceNo StockCode                          Description  Quantity  \
142914    548654     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
142915    548654     22055   MINI CAKE STAND  HANGING STRAWBERY        16   
142916    548654     22060      LARGE CAKE STAND HANGING HEARTS        16   
142917    548654     21700          BIG DOUGHNUT FRIDGE MAGNETS        12   
142918    548654     37447    CERAMIC CAKE DESIGN SPOTTED PLATE        12   
...          ...       ...                                  ...       ...   
383687    570082      POST                              POSTAGE         7   
515713    579786     22326  ROUND SNACK BOXES SET OF4 WOODLAND         48   
515714    579786     21580      RABBIT  DESIGN  COTTON TOTE BAG        12   
515715    579786     22327    ROUND SNACK BOXES SET OF 4 SKULLS        48   
515716    579786      POST                              POSTAGE         2   

             InvoiceDate  UnitPrice  CustomerID  Country  
142914    4/1/2011 13:47       1.25     12569.0  Germany  
142915    4/1/2011 13:47       1.65     12569.0  Germany  
142916    4/1/2011 13:47       3.95     12569.0  Germany  
142917    4/1/2011 13:47       0.85     12569.0  Germany  
142918    4/1/2011 13:47       1.49     12569.0  Germany  
...                  ...        ...         ...      ...  
383687   10/7/2011 11:19      18.00     12524.0  Germany  
515713  11/30/2011 15:31       2.55     12524.0  Germany  
515714  11/30/2011 15:31       2.25     12524.0  Germany  
515715  11/30/2011 15:31       2.55     12524.0  Germany  
515716  11/30/2011 15:31      18.00     12524.0  Germany  

[4175 rows x 8 columns]

# Сгруппируйте top_retail_germany по коду товара (StockCode).   

Какой из продуктов добавляли в корзину чаще всего, кроме POST?   
Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.

In [219]:
top_retail_germany.groupby('StockCode') \
    .agg({'InvoiceNo':'count'}) \
    .sort_values('InvoiceNo',ascending=False) \
    .rename(columns = {'InvoiceNo':'Count'})

Count
StockCode       
POST         213
22326         52
22328         38
22423         34
20719         30
...          ...
22563          1
22569          1
22574          1
22576          1
22662          1

[1158 rows x 1 columns]

Чаще всего в корзину добавляли продукт 22326

# Вернемся к анализу полного датасета retail.   
Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.

In [227]:
value = retail['Quantity'] * retail['UnitPrice']
retail['Revenue'] = value

<ipython-input-227-b319a463ae51>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  retail['Revenue'] = value


In [228]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
...                 ...        ...         ...             ...      ...  
541904  12/9/2011 12:50       0.85     12680.0          France    10.20  
541905  12/9/2011 12:50       2.10     12680.0          France    12.60  
541906  12/9/2011 12:50       4.15     12680.0          France    16.60  
541907  12/9/2011 12:50       4.15     12680.0          France    16.60  
541908  12/9/2011 12:50       4.95     12680.0          France    14.85  

[526054 rows x 9 columns]

# Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа.  
В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)


In [236]:
top_5_rev = retail.groupby('InvoiceNo') \
    .agg({'Revenue':'sum'}) \
    .sort_values('Revenue',ascending=False) \
    .rename(columns={'Revenue':'total_cost'}) \
    .head(5)

In [249]:
lst = top_5_rev.index
lst.str.cat(sep=', ')

'581483, 541431, 574941, 576365, 556444'